### PR TITLE
cli: Enable cockroach debug tsdump to automatically create tsdump.yaml

### PR DESCRIPTION
**Pull request by daniel-crlabs**
Release note (cli change): cockroach debug tsdump creates tsdump.yaml

The changes provided by this PR will enable users when creating a tsdump in raw format, to automatically create a tsdump.yaml file containing the storeID: nodeID mappings. This file is required by Technical Support in order to be able to load tsdump raw files for troubleshooting purposes, so this feature will automate the process of creating such yaml file.

This change will expedite the TSEs ability to investigate issues requiring the tsdump, as it eliminates the
requirement for TSEs to also obtain the debug zip in order to create the tsdump.yaml file.

The tsdump.yaml is only created automatically if the flag `--format=raw` is used, if the user choses any other format the yaml file will not be created. The tsdump.yaml will be automatically created in `/tmp/tsump.yaml`.

In addition to using the `--format=raw` flag, an user can pass the `--yaml` flag, allowing the user to chose where to save the yaml file, instead of the default location (`/tmp`).

Usage:

```
cockroach debug tsdump --host <host>:<port> --format raw --yaml=/some_path/tsdump.yaml > /some_path/tsdump.gob
```

Authored by: Daniel Almeida

### CODE REVIEW

**Comment by daniel-crlabs** [comment]
I happened to get one of those automated emails about a stale issue, and long and behold, it was for a request made 2 years ago to have this feature possible. I'm adding it here for reference purposes to show this is not the first time this has come up, as the changes I have created in this PR will solve exactly this problem and fulfill this requirement.

[cli: tsdump should also create tsdump.gob.yaml #80478](https://github.com/cockroachdb/cockroach/issues/80478)

**Review by dhartunian — COMMENTED** [review]
Thanks for filing this PR!

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-:-Nj9rESC3nLlAqfAkvvA:b-8evcyt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @daniel-crlabs)

___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9qu_Q9BlhmAPTnlpf):*
nit: `tsdump` in path.
___
*[`-- commits` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9s6My2lq2mU_BBkj6):*
you should add more detail for the docs team here. I think two things are worth highlighting (and will be repeated from text above):
1. raw format automatically creates yaml
2. new flag `--yaml` that controls output filename
___
*[`pkg/cli/tsdump.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9r29o0T2LeUlI44XI:-Nj9r29o0T2LeUlI44XJ:b6qglwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L26)):*
> ```Go
> 
> 	"github.com/cockroachdb/cockroach/pkg/cli/clierrorplus"
> 	"github.com/cockroachdb/cockroach/pkg/cli/clisqlclient" // Added by Daniel Almeida to enable yaml to be created
> ```

no need for this comment. information about who and why someone added a dependency can be gleaned from the git history and your commit message already.
___
*[`pkg/cli/tsdump.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9r9RRCIbrM-A0OoqS:-Nj9r9RRCIbrM-A0OoqT:b-br8jjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L43)):*
> ```Go
> 	from, to     timestampValue
> 	clusterLabel string
> 	yaml         string // Added by Daniel Almeida to enable yaml to be created
> ```

no need for this comment
___
*[`pkg/cli/tsdump.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9rAu_1DEa5koSeZwC:-Nj9rAu_1DEa5koSeZwD:b8a0kqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L49)):*
> ```Go
> 	to:           timestampValue(timeutil.Now().Add(24 * time.Hour)),
> 	clusterLabel: "",
> 	yaml:         "/tmp/tsdump.yaml", // Added by Daniel Almeida to enable yaml to be created. Default location is /tmp/tsdump.yaml
> ```

ditto. no need for this comment. 
___
*[`pkg/cli/tsdump.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9rg5s3DXx4tFVkxr2:-Nj9rg5s3DXx4tFVkxr3:bd13wx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L214)):*
> ```Go
> 
> // Daniel Almeida function to obtain storeID to nodeID mappings when creating raw tsdump
> // We automatically create the tsdump.yaml whenever creating a raw tsdump, as this is required for loading raw time series data
> ```

The format for function docstrings is detailed here: https://go.dev/blog/godoc

So in this case you would write:
```
// createYAML generates and writes the yaml to...
```
___
*[`pkg/cli/tsdump.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-NjA-BJS-0kLKNDX7TB6:-NjA-BJS-0kLKNDX7TB7:b9nlv21) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L222)):*
> ```Go
> 	defer file.Close()
> 
> 	ctx := context.Background()
> ```

you can pass in the `ctx` from the caller of `createYAML` instead of making a new one here.
___
*[`pkg/cli/tsdump.go` line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9zYEC2wu56L8QyLar:-Nj9zYED71rHk_Rj1Aec:b-a1n8ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L223)):*
> ```Go
> 
> 	ctx := context.Background()
> 	sqlConn, err := makeSQLClient(ctx, "tsdump-node-to-store-mapping", useSystemDb)
> ```

to follow patterns in rest of the `pkg/cli` codebase I would use `cockroach tsdump` as the appname here.

**Comment by daniel-crlabs** [comment]
> Thanks for filing this PR!
> 
> > Reviewed 1 of 2 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-:-Nj9rESC3nLlAqfAkvvA:b-8evcyt)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @daniel-crlabs)
> 
> _[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9qu_Q9BlhmAPTnlpf):_ nit: `tsdump` in path.
> 
> _[`-- commits` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9s6My2lq2mU_BBkj6):_ you should add more detail for the docs team here. I think two things are worth highlighting (and will be repeated from text above):
> 
> 1. raw format automatically creates yaml
> 2. new flag `--yaml` that controls output filename
> 
> _[`pkg/cli/tsdump.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9r29o0T2LeUlI44XI:-Nj9r29o0T2LeUlI44XJ:b6qglwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L26)):_
> 
> > ```go
> > 	"github.com/cockroachdb/cockroach/pkg/cli/clierrorplus"
> > 	"github.com/cockroachdb/cockroach/pkg/cli/clisqlclient" // Added by Daniel Almeida to enable yaml to be created
> > ```
> 
> no need for this comment. information about who and why someone added a dependency can be gleaned from the git history and your commit message already.
> 
> _[`pkg/cli/tsdump.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9r9RRCIbrM-A0OoqS:-Nj9r9RRCIbrM-A0OoqT:b-br8jjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L43)):_
> 
> > ```go
> > 	from, to     timestampValue
> > 	clusterLabel string
> > 	yaml         string // Added by Daniel Almeida to enable yaml to be created
> > ```
> 
> no need for this comment
> 
> _[`pkg/cli/tsdump.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9rAu_1DEa5koSeZwC:-Nj9rAu_1DEa5koSeZwD:b8a0kqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L49)):_
> 
> > ```go
> > 	to:           timestampValue(timeutil.Now().Add(24 * time.Hour)),
> > 	clusterLabel: "",
> > 	yaml:         "/tmp/tsdump.yaml", // Added by Daniel Almeida to enable yaml to be created. Default location is /tmp/tsdump.yaml
> > ```
> 
> ditto. no need for this comment.
> 
> _[`pkg/cli/tsdump.go` line 214 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9rg5s3DXx4tFVkxr2:-Nj9rg5s3DXx4tFVkxr3:bd13wx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L214)):_
> 
> > ```go
> > // Daniel Almeida function to obtain storeID to nodeID mappings when creating raw tsdump
> > // We automatically create the tsdump.yaml whenever creating a raw tsdump, as this is required for loading raw time series data
> > ```
> 
> The format for function docstrings is detailed here: https://go.dev/blog/godoc
> 
> So in this case you would write:
> 
> ```
> // createYAML generates and writes the yaml to...
> ```
> 
> _[`pkg/cli/tsdump.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-NjA-BJS-0kLKNDX7TB6:-NjA-BJS-0kLKNDX7TB7:b9nlv21) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L222)):_
> 
> > ```go
> > 	defer file.Close()
> > 
> > 	ctx := context.Background()
> > ```
> 
> you can pass in the `ctx` from the caller of `createYAML` instead of making a new one here.
> 
> _[`pkg/cli/tsdump.go` line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-Nj9zYEC2wu56L8QyLar:-Nj9zYED71rHk_Rj1Aec:b-a1n8ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/a56dc25282b6f257afdf56f9d7858ea352acd3b5/pkg/cli/tsdump.go#L223)):_
> 
> > ```go
> > 	ctx := context.Background()
> > 	sqlConn, err := makeSQLClient(ctx, "tsdump-node-to-store-mapping", useSystemDb)
> > ```
> 
> to follow patterns in rest of the `pkg/cli` codebase I would use `cockroach tsdump` as the appname here.

Sorry I forgot to comment on the ticket (I thought posting the updates to the code would trigger a new review automatically). Please review when you have a chance, I've done all the changes.

**Review by dhartunian — APPROVED** [review]
:lgtm: 

@daniel-crlabs now that this PR is approved, if you post a github comment with the text`bors r+` this PR will enter the merge queue and end up in the `master` branch.

Backport PRs will automatically be created by the bot and I will approve those once they're up as well. Backport PRs can be merged after approval with the standard GitHub merge button (no bors required)

Also, please open a PR to add your name to the `AUTHORS` file :tada: (https://github.com/cockroachdb/cockroach/blob/master/AUTHORS)

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114046#-:-NkbQc6_C5D_QnpI21Ap:b-f1o9yk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @daniel-crlabs)

**Comment by daniel-crlabs** [comment]
Thank you, appreciate the work on this.

bors r+

**Comment by daniel-crlabs** [comment]
Hi @dhartunian I'm a bit confused by all of the errors above, is this a concern?

**Comment by mattcrdb** [comment]
ping @dhartunian

**Comment by dhartunian** [comment]
Backport for 23.2 is up. 23.1 is a lot more tricky. Can we get by with the 23.2 backport alone?

**Comment by mattcrdb** [comment]
We'll need 23.1 and 22.2 (6 mo support remaining) as well. Customers will choose to stay one version back for a major version so backporting to 23.2 will leave us little benefit for this change in the short term.

**Comment by nickelnickel** [comment]
To @mattcrdb's point we need 23.1. Having to collect debug.zip in addition to a tsdump is a major pain point with a number of customers.

**Pull request merged**
